src/commonMain/kotlin/fr/acinq/lightning/blockchain/electrum/ElectrumMiniWallet.kt

```diff
@@ -24,20 +24,31 @@ data class WalletState(val addresses: Map<String, List<UnspentItem>>, val parent
         .filter { parentTxs.containsKey(it.txid) }
         .map { Utxo(parentTxs[it.txid]!!, it.outputIndex, it.blockHeight) }
 
-    val confirmedBalance = balance(includingUnconfirmed = false, includingConfirmed = true)
-    val unconfirmedBalance = balance(includingUnconfirmed = true, includingConfirmed = false)
+    val confirmedUtxos: List<Utxo> = utxos.filter { it.blockHeight > 0L }
+    val unconfirmedUtxos: List<Utxo> = utxos.filter { it.blockHeight == 0L }
+
+    val confirmedBalance = confirmedUtxos.map { it.amount }.sum()
+    val unconfirmedBalance = unconfirmedUtxos.map { it.amount }.sum()
     val totalBalance = confirmedBalance + unconfirmedBalance
-    private fun balance(includingUnconfirmed: Boolean, includingConfirmed: Boolean): Satoshi =
-        utxos
-            .filter { (includingUnconfirmed && it.blockHeight == 0L) || (includingConfirmed && it.blockHeight > 0L) }
-            .map { it.amount }
-            .sum()
+
+    fun drop(reserved: Set<Utxo>): WalletState {
```

Comment:
> due to the structure of WalletState which has been designed to work easily with electrum

I thought the structure of WalletState was quite clever, given that it's trying to deal with **asynchronous** Electrum requests. It does what it must, in a rather readable form, and presents the usual suspects to external users (`List<Utxo>` & `balance`)

> I'm not 100% happy with the idea of creating subsets of the `WalletState` class. In fact, we only really need a `List<Utxo>`

Exactly. That was my first idea, but then when I started implementing it I ran into this problem:

> Unfortunately this has impacts on the rest of the codebase and going with this would clash with #391

Yup. A bunch of classes explicitly take a `WalletState` parameter. Given the new approach (of reserving Utxos) it would make more sense to refactor these classes, and instead pass `List<Utxo>`. But I agree with you, it's probably better to revisit this later, after all the conflicting in-flight PRs are closed.